File: backend/app/schemas/scan.py
"""Pydantic mirror of the Aegis forensic scan pipeline contract.

This module is the BACKEND HALF of the four-way contract mirror. The other
three are:

  * `contract/README.md`              — human-readable spec
  * `contract/scan-pipeline.schema.json` — language-neutral JSON Schema
  * `src/lib/contract/scan.ts`        — frontend TypeScript types

A shape change here MUST land in all four at once. The Pydantic models are
configured with `populate_by_name=True` and camelCase aliases so the wire
JSON is byte-compatible with the TS contract.
"""
from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

# ─── Base config ──────────────────────────────────────────────────────────


class ContractModel(BaseModel):
    """Every contract DTO renders camelCase on the wire."""

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        str_strip_whitespace=True,
        extra="forbid",
        frozen=True,
    )


# ─── Primitives ───────────────────────────────────────────────────────────

Verdict = Literal["PASS", "REVIEW", "FAIL"]
VlmSeverity = Literal["info", "warning", "critical"]
ScanPipelineStage = Literal[
    "normalization_complete",
    "identify_complete",
    "ela_complete",
    "vlm_complete",
    "consensus",
]


class ElaRect(ContractModel):
    """A single ELA hot-spot in normalised (0..1) coordinates."""

    x: Annotated[float, Field(ge=0.0, le=1.0)]
    y: Annotated[float, Field(ge=0.0, le=1.0)]
    w: Annotated[float, Field(ge=0.0, le=1.0)]
    h: Annotated[float, Field(ge=0.0, le=1.0)]
    intensity: Annotated[float, Field(ge=0.0, le=1.0)]


class VlmFinding(ContractModel):
    """One print-discrepancy finding."""

    id: Annotated[int, Field(ge=1)]
    x: Annotated[float, Field(ge=0.0, le=1.0)]
    y: Annotated[float, Field(ge=0.0, le=1.0)]
    title: str
    detail: str
    severity: VlmSeverity


# ─── Stage payloads ───────────────────────────────────────────────────────


class NormalizationPayload(ContractModel):
    session_id: str
    product_name: str
    bbox: Annotated[
        tuple[float, float, float, float],
        Field(description="Detected label bbox normalised [x, y, w, h]"),
    ]
    normalized_image_url: str


class IdentifyPayload(ContractModel):
    """Stage 8 — Claude vision classified the pack."""

    session_id: str
    product_name: str
    nafdac_reg_number: str | None = None
    confidence: Annotated[float, Field(ge=0.0, le=1.0)]
    reference_key: str | None = None


class ElaPayload(ContractModel):
    session_id: str
    product_name: str
    ela_score: Annotated[float, Field(ge=0.0, le=100.0)]
    ela_map: list[ElaRect]
    heatmap_image_url: str | None = None


class VlmPayload(ContractModel):
    session_id: str
    product_name: str
    # `None` when no reference was available for the identified product.
    vlm_score: float | None = Field(default=None, ge=0.0, le=100.0)
    vlm_findings: list[VlmFinding] = Field(default_factory=list)
    reference_image: str | None = None
    reference_reg_number: str | None = None
    reference_available: bool = True


class ScanResult(ContractModel):
    """Terminal consensus payload — also returned by GET /api/scan/{id}/result."""

    session_id: str
    product_name: str
    batch_id: str | None = None
    captured_at: str
    ela_score: Annotated[float, Field(ge=0.0, le=100.0)]
    # `None` when no reference was available — verdict relies on ELA alone.
    vlm_score: float | None = Field(default=None, ge=0.0, le=100.0)
    consensus_score: Annotated[float, Field(ge=0.0, le=100.0)]
    verdict: Verdict
    ela_map: list[ElaRect]
    vlm_findings: list[VlmFinding]
    summary: str
    reference_image: str | None = None
    reference_available: bool = True


# ─── Pipeline events (discriminated union on `stage`) ─────────────────────


class _BaseEvent(ContractModel):
    timestamp: str
    latency_ms: Annotated[int, Field(ge=0)]


class NormalizationCompleteEvent(_BaseEvent):
    stage: Literal["normalization_complete"] = "normalization_complete"
    payload: NormalizationPayload


class IdentifyCompleteEvent(_BaseEvent):
    stage: Literal["identify_complete"] = "identify_complete"
    payload: IdentifyPayload


class ElaCompleteEvent(_BaseEvent):
    stage: Literal["ela_complete"] = "ela_complete"
    payload: ElaPayload


class VlmCompleteEvent(_BaseEvent):
    stage: Literal["vlm_complete"] = "vlm_complete"
    payload: VlmPayload


class ConsensusEvent(_BaseEvent):
    stage: Literal["consensus"] = "consensus"
    payload: ScanResult


ScanPipelineEvent = Annotated[
    NormalizationCompleteEvent
    | IdentifyCompleteEvent
    | ElaCompleteEvent
    | VlmCompleteEvent
    | ConsensusEvent,
    Field(discriminator="stage"),
]


class ScanErrorEvent(ContractModel):
    stage: Literal["error"] = "error"
    timestamp: str
    failed_stage: ScanPipelineStage | Literal["upload"]
    message: str
    recoverable: bool


# ─── HTTP DTOs ────────────────────────────────────────────────────────────


class ScanAcceptResponse(ContractModel):
    session_id: str
    accepted_at: str


__all__ = [
    "ConsensusEvent",
    "ContractModel",
    "ElaCompleteEvent",
    "ElaPayload",
    "ElaRect",
    "IdentifyCompleteEvent",
    "IdentifyPayload",
    "NormalizationCompleteEvent",
    "NormalizationPayload",
    "ScanAcceptResponse",
    "ScanErrorEvent",
    "ScanPipelineEvent",
    "ScanPipelineStage",
    "ScanResult",
    "Verdict",
    "VlmCompleteEvent",
    "VlmFinding",
    "VlmPayload",
    "VlmSeverity",
]

File: src/lib/contract/scan.ts
/**
 * ╔══════════════════════════════════════════════════════════════════════╗
 * ║  AEGIS — FORENSIC SCAN PIPELINE CONTRACT                              ║
 * ║  Single source of truth for the scan pipeline wire format.           ║
 * ╚══════════════════════════════════════════════════════════════════════╝
 *
 * This file is the canonical TypeScript definition of the contract between
 * the Next.js frontend and the FastAPI forensic backend. The backend mirrors
 * these shapes in `backend/app/schemas/scan.py` (Pydantic), and the
 * human-readable spec lives in `/contract/README.md` with a machine-readable
 * JSON Schema in `/contract/scan-pipeline.schema.json`.
 *
 * RULE: if a shape changes here, it changes in all four places, in lockstep.
 *
 * Transport: the pipeline streams over Server-Sent Events (SSE). Each SSE
 * message carries `event: <stage>` and `data: <JSON of ScanPipelineEvent>`.
 * See `/contract/README.md` §Transport for the rationale.
 */

// ─────────────────────────────────────────────────────────────────────────
// Primitives
// ─────────────────────────────────────────────────────────────────────────

/** Final forensic verdict. >=85 PASS · 60–84 REVIEW · <60 FAIL. */
export type Verdict = "PASS" | "REVIEW" | "FAIL";

/** Severity of a single VLM print-discrepancy finding. */
export type VlmSeverity = "info" | "warning" | "critical";

/**
 * A single ELA hot-spot, in normalised (0..1) coordinates relative to the
 * normalised capture. `intensity` (0..1) drives the yellow→red gradient.
 */
export interface ElaRect {
  x: number;
  y: number;
  w: number;
  h: number;
  intensity: number;
}

/**
 * A single visual discrepancy the VLM stage found against the NAFDAC
 * reference. `x`/`y` are normalised (0..1) marker positions on the capture.
 */
export interface VlmFinding {
  /** 1-indexed marker number, stable within a single scan. */
  id: number;
  x: number;
  y: number;
  title: string;
  detail: string;
  severity: VlmSeverity;
}

// ─────────────────────────────────────────────────────────────────────────
// Pipeline stages
// ─────────────────────────────────────────────────────────────────────────

/**
 * Pipeline stages.
 *
 * `identify_complete` is conditionally emitted — it only fires when the
 * scan was accepted without batch-supplied product context (see D-0028).
 * The other four stages always fire, in the order below.
 */
export type ScanPipelineStage =
  | "normalization_complete"
  | "identify_complete"
  | "ela_complete"
  | "vlm_complete"
  | "consensus";

export const SCAN_PIPELINE_STAGE_ORDER: readonly ScanPipelineStage[] = [
  "normalization_complete",
  "identify_complete",
  "ela_complete",
  "vlm_complete",
  "consensus",
] as const;

// ─────────────────────────────────────────────────────────────────────────
// Stage payloads
// ─────────────────────────────────────────────────────────────────────────

/**
 * `normalization_complete` — the CNN/OpenCV stage detected the label,
 * computed the homography, and warped the capture flat.
 */
export interface NormalizationPayload {
  sessionId: string;
  productName: string;
  /** Detected label bounding box, normalised [x, y, w, h]. */
  bbox: [number, number, number, number];
  /** URL to the perspective-corrected capture. */
  normalizedImageUrl: string;
}

/**
 * `identify_complete` — Claude vision classified the pack and returned a
 * product name + optional NAFDAC reg number. Only fires for sessions
 * without batch-supplied product context.
 *
 * `referenceKey` is the backend's catalogue key (e.g. "coartem",
 * "augmentin") when the identified product matches one of the references
 * the VLM can compare against. `null` means no reference available — the
 * downstream VLM stage degrades to identification-only (vlmScore=null on
 * the wire, vlmFindings empty), and consensus drops VLM's weight to zero.
 */
export interface IdentifyPayload {
  sessionId: string;
  productName: string;
  /** NAFDAC registration number Claude read off the pack, if present. */
  nafdacRegNumber: string | null;
  /** 0..1 self-reported identification confidence. */
  confidence: number;
  /**
   * Backend catalogue key for the matched reference image. `null` when the
   * identified product has no reference on file.
   */
  referenceKey: string | null;
}

/**
 * `ela_complete` — Error Level Analysis produced a digital-integrity score
 * and a set of hot-spot rectangles.
 */
export interface ElaPayload {
  sessionId: string;
  productName: string;
  /** 0..100 — higher is cleaner. */
  elaScore: number;
  elaMap: ElaRect[];
  /** Optional raster heatmap, if the backend chooses to ship one. */
  heatmapImageUrl?: string;
}

/**
 * `vlm_complete` — the Claude vision stage compared the normalised capture
 * against the NAFDAC golden reference and returned print-integrity findings.
 */
export interface VlmPayload {
  sessionId: string;
  productName: string;
  /**
   * 0..100 — higher is cleaner. `null` when no NAFDAC reference was
   * available for the identified product (consumer scanned a pack we
   * don't have a golden image for); in that case `vlmFindings` is `[]`
   * and consensus drops VLM's weight to zero.
   */
  vlmScore: number | null;
  vlmFindings: VlmFinding[];
  /** URL to the NAFDAC reference image used for the comparison. `null` when no reference. */
  referenceImage: string | null;
  /** NAFDAC registration number the reference was keyed by. `null` when no reference. */
  referenceRegNumber: string | null;
  /** True when a real reference comparison ran; false for identify-only. */
  referenceAvailable: boolean;
}

/**
 * `consensus` — the terminal payload. The weighted verdict plus every score
 * and artefact the verdict screen needs to render the final card. This is
 * the canonical wire shape; the mock catalogue's `ScanFixture` is a superset
 * of this used only by the in-browser simulator.
 */
export interface ScanResult {
  /** The scan session id minted by `POST /api/scan`. */
  sessionId: string;
  productName: string;
  /** Batch the scan was performed in support of, if any. */
  batchId?: string;
  /** ISO 8601 capture timestamp. */
  capturedAt: string;
  elaScore: number;
  /** `null` when no NAFDAC reference was available — see VlmPayload. */
  vlmScore: number | null;
  /** Weighted average of ELA + VLM (ELA-only when vlmScore is null). */
  consensusScore: number;
  verdict: Verdict;
  elaMap: ElaRect[];
  vlmFindings: VlmFinding[];
  /** One-line summary shown above the verdict card. */
  summary: string;
  /** `null` when no NAFDAC reference was available. */
  referenceImage: string | null;
  /** True when the VLM compared against a real reference. */
  referenceAvailable: boolean;
}

// ─────────────────────────────────────────────────────────────────────────
// Pipeline events (discriminated union on `stage`)
// ─────────────────────────────────────────────────────────────────────────

interface BasePipelineEvent {
  /** ISO 8601 timestamp the stage completed. */
  timestamp: string;
  /** Wall-clock milliseconds this stage took. */
  latencyMs: number;
}

export interface NormalizationCompleteEvent extends BasePipelineEvent {
  stage: "normalization_complete";
  payload: NormalizationPayload;
}

export interface IdentifyCompleteEvent extends BasePipelineEvent {
  stage: "identify_complete";
  payload: IdentifyPayload;
}

export interface ElaCompleteEvent extends BasePipelineEvent {
  stage: "ela_complete";
  payload: ElaPayload;
}

export interface VlmCompleteEvent extends BasePipelineEvent {
  stage: "vlm_complete";
  payload: VlmPayload;
}

export interface ConsensusEvent extends BasePipelineEvent {
  stage: "consensus";
  payload: ScanResult;
}

/** Any event the pipeline SSE stream can emit. */
export type ScanPipelineEvent =
  | NormalizationCompleteEvent
  | IdentifyCompleteEvent
  | ElaCompleteEvent
  | VlmCompleteEvent
  | ConsensusEvent;

/**
 * Emitted on the SSE stream's `error` event channel when a stage fails. The
 * client should fall back to `GET /api/scan/:id/result` when `recoverable`.
 */
export interface ScanErrorEvent {
  stage: "error";
  timestamp: string;
  /** The stage that was in flight when the failure occurred. */
  failedStage: ScanPipelineStage | "upload";
  message: string;
  recoverable: boolean;
}

// ─────────────────────────────────────────────────────────────────────────
// HTTP DTOs
// ─────────────────────────────────────────────────────────────────────────

/** Response to `POST /api/scan` (multipart capture upload). */
export interface ScanAcceptResponse {
  sessionId: string;
  /** ISO 8601 timestamp the capture was accepted. */
  acceptedAt: string;
}

/** Response to `GET /api/health`. */
export interface HealthResponse {
  status: "ok";
  service: string;
  version: string;
  env: string;
  time: string;
}

// ─────────────────────────────────────────────────────────────────────────
// Narrowing helpers
// ─────────────────────────────────────────────────────────────────────────

export function isConsensusEvent(
  e: ScanPipelineEvent
): e is ConsensusEvent {
  return e.stage === "consensus";
}

export function isElaEvent(e: ScanPipelineEvent): e is ElaCompleteEvent {
  return e.stage === "ela_complete";
}

export function isVlmEvent(e: ScanPipelineEvent): e is VlmCompleteEvent {
  return e.stage === "vlm_complete";
}

export function isNormalizationEvent(
  e: ScanPipelineEvent
): e is NormalizationCompleteEvent {
  return e.stage === "normalization_complete";
}

export function isIdentifyEvent(
  e: ScanPipelineEvent
): e is IdentifyCompleteEvent {
  return e.stage === "identify_complete";
}
